[FSDP] Default `limit_all_gathers=True`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104900

This PR defaults to `limit_all_gathers=True`.

I included a `record_function()` for the rate limiter synchronization to help with user confusion on the gap in the pre-forward:
<img width="874" alt="Screenshot 2023-07-10 at 3 28 18 PM" src="https://github.com/pytorch/pytorch/assets/31054793/61f55e0e-58d7-4162-9395-bea06d3e8d8a">

Commit:
[FSDP] Default `limit_all_gathers=True`

[ghstack-poisoned]

fegin reviewed (approved):
LGTM, but should we wait for the internal diffs are landed before this PR is landed?

fduwjj reviewed (commented):
Nice!!

Commit:
Update on "[FSDP] Default `limit_all_gathers=True`"


This PR defaults to `limit_all_gathers=True`.

I included a `record_function()` for the rate limiter synchronization to help with user confusion on the gap in the pre-forward:
<img width="874" alt="Screenshot 2023-07-10 at 3 28 18 PM" src="https://github.com/pytorch/pytorch/assets/31054793/61f55e0e-58d7-4162-9395-bea06d3e8d8a">


[ghstack-poisoned]

awgu commented:
Internal config diff is landing.

awgu commented:
@pytorchbot merge
